File: ios/Classes/NeoVapPlugin.swift
import Flutter
import UIKit

/// Texture-based transparent VAP player. Owns the method/event channels and a
/// map of live [NeoVapPlayer]s keyed by texture id (the iOS mirror of
/// `NeoVapPlugin.kt`). Decode + composite happen per-player (AVPlayer -> Metal
/// -> FlutterTexture); this class is just wiring.
public class NeoVapPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private let registrar: FlutterPluginRegistrar
  private let textures: FlutterTextureRegistry
  private var players = [Int64: NeoVapPlayer]()
  private var eventSink: FlutterEventSink?

  init(registrar: FlutterPluginRegistrar) {
    self.registrar = registrar
    self.textures = registrar.textures()
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = NeoVapPlugin(registrar: registrar)
    let channel = FlutterMethodChannel(name: "neo_vap", binaryMessenger: registrar.messenger())
    registrar.addMethodCallDelegate(instance, channel: channel)
    let events = FlutterEventChannel(name: "neo_vap/events", binaryMessenger: registrar.messenger())
    events.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    switch call.method {
    case "allocateTexture":
      let player = NeoVapPlayer(registry: textures)
      let id = textures.register(player)
      player.textureId = id
      player.onEvent = { [weak self] event, message in
        self?.emit(textureId: id, event: event, message: message)
      }
      players[id] = player
      result(id)

    case "play":
      guard
        let args = args, let id = int64(args["textureId"]),
        let asset = args["asset"] as? String, let url = assetURL(asset)
      else { return result(argError("play")) }
      let next = (args["nextAsset"] as? String).flatMap(assetURL)
      players[id]?.play(
        assetURL: url,
        vapcSource: url,
        repeatMode: (args["repeat"] as? Int) ?? NeoVapPlayer.loopForever,
        nextURL: next
      )
      result(nil)

    case "stop":
      players[textureId(args)]?.stop()
      result(nil)

    case "dispose":
      let id = textureId(args)
      players.removeValue(forKey: id)?.dispose()
      textures.unregisterTexture(id)
      result(nil)

    case "prewarm":
      prewarm()
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Warm the Metal device + composite pipeline once per process so the first
  /// real play doesn't pay the cold PSO/shader-compile stall (the iOS latency
  /// fix). Runs off the main thread; best-effort, so any failure just means the
  /// first play pays full cold-start.
  ///
  /// The three prewarm-hardening lessons carried from Android:
  ///  1. All construction is inside `MetalCompositor.init?` (returns nil, never
  ///     throws) so a build failure on this background thread can't crash init.
  ///  2. The Dart facade is fire-and-forget (`.catchError`), so having a real
  ///     handler here just means it succeeds instead of MissingPlugin-rejecting.
  ///  3. No blocking latch on the main thread — this dispatches async and returns
  ///     immediately; nothing the main thread waits on can wedge it (no ANR).
  private func prewarm() {
    DispatchQueue.global(qos: .utility).async {
      if MetalCompositor.shared != nil {
        NSLog("neo_vap: prewarm — Metal composite pipeline warmed")
      } else {
        NSLog("neo_vap: prewarm skipped (Metal unavailable)")
      }
    }
  }

  // MARK: - helpers

  private func assetURL(_ asset: String) -> URL? {
    let key = registrar.lookupKey(forAsset: asset)
    guard let path = Bundle.main.path(forResource: key, ofType: nil) else { return nil }
    return URL(fileURLWithPath: path)
  }

  private func int64(_ any: Any?) -> Int64? { (any as? NSNumber)?.int64Value }

  private func textureId(_ args: [String: Any]?) -> Int64 {
    int64(args?["textureId"]) ?? -1
  }

  private func argError(_ method: String) -> FlutterError {
    FlutterError(code: "neo_vap", message: "bad arguments for \(method)", details: nil)
  }

  private func emit(textureId: Int64, event: String, message: String?) {
    // NSNull (not a Swift nil-in-dict, which bridges to the codec badly) so Dart
    // reads `message` back as null. Matches the Android event map shape.
    let payload: [String: Any] = [
      "textureId": textureId, "event": event, "message": message ?? NSNull(),
    ]
    DispatchQueue.main.async { [weak self] in self?.eventSink?(payload) }
  }

  // MARK: - FlutterStreamHandler

  public func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    return nil
  }

  public func onCancel(withArguments _: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  public func detachFromEngine(for _: FlutterPluginRegistrar) {
    players.values.forEach { $0.dispose() }
    players.removeAll()
    eventSink = nil
  }
}

File: android/src/main/kotlin/xyz/neosapien/neo_vap/NeoVapPlugin.kt
package xyz.neosapien.neo_vap

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry

/**
 * Texture-based transparent VAP player. Owns the method/event channels and a
 * map of live [NeoVapPlayer]s keyed by texture id. Decode + composite happen
 * per-player (ExoPlayer -> GL -> SurfaceProducer); this class is just wiring.
 */
class NeoVapPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private lateinit var textureRegistry: TextureRegistry

    private val players = mutableMapOf<Long, NeoVapPlayer>()
    private val main = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        textureRegistry = binding.textureRegistry
        methodChannel = MethodChannel(binding.binaryMessenger, "neo_vap")
        methodChannel.setMethodCallHandler(this)
        eventChannel = EventChannel(binding.binaryMessenger, "neo_vap/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "allocateTexture" -> {
                    val producer = textureRegistry.createSurfaceProducer()
                    val id = producer.id()
                    players[id] = NeoVapPlayer(context, producer) { event, message ->
                        emit(id, event, message)
                    }
                    result.success(id)
                }
                "play" -> {
                    player(call)?.play(
                        call.arg("asset"),
                        call.argument<Int>("repeat") ?: NeoVapPlayer.LOOP_FOREVER,
                        call.argument<String>("nextAsset"),
                    )
                    result.success(null)
                }
                "stop" -> {
                    player(call)?.stop()
                    result.success(null)
                }
                "dispose" -> {
                    players.remove(textureId(call))?.dispose()
                    result.success(null)
                }
                "prewarm" -> {
                    prewarm()
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("neo_vap", e.message, null)
        }
    }

    /**
     * Warm the GL driver + shader compiler once per process so the first real
     * `play` doesn't pay the cold EGL-init + program-compile stall. Runs off the
     * main thread; best-effort, so any failure just means the first play pays
     * full cold-start.
     *
     * Measured on a vivo V2521 (API 35, Adreno, c2.qti.avc.decoder), profile
     * build, cold start: native init (asset read + vapc parse + GL + ExoPlayer
     * construction) is 38-71ms total, of which GL is ~3-27ms. The first real
     * decoded frame lands at t+313-334ms — i.e. init is only ~15% of
     * time-to-content. The other ~260ms is downstream of `prepare()`: MediaSource
     * load and extractor parse (~156ms) then codec init and first decode (~61ms).
     * Optimising init further has little left to win; that ~260ms is the target.
     *
     * Deliberately NOT warmed here: `MediaCodecUtil.warmDecoderInfoCache`.
     * `getDecoderInfos` is `static synchronized`, so warming it on this thread
     * does not delete the cost for a play starting moments later — the play just
     * blocks on the same monitor. It also enumerates the whole MediaCodecList,
     * which is itself the source of the ~30 AudioCapabilities/VideoCapabilities
     * "Unsupported mime" warnings (verified on tid neo_vap_prewarm) — so it adds
     * the log spam it was assumed to avoid. Revisit only with real runway ahead
     * of the first play.
     */
    private fun prewarm() {
        if (warmed) return
        warmed = true
        Thread {
            // 16x16 throwaway geometry: enough to run initGl (EGL init + OES
            // texture + shader compile/link), which is all we're warming.
            val dummy = VapcInfo(
                version = 2, frameCount = 1, width = 16, height = 16, fps = 25,
                videoWidth = 16, videoHeight = 16,
                rgbFrame = VapcRect(0, 0, 16, 16), aFrame = VapcRect(0, 0, 16, 16),
                isVapx = false, orientation = 0,
            )
            val start = SystemClock.elapsedRealtime()
            // Construct inside the try: NeoVapRenderer() starts a HandlerThread, so
            // a thread-creation/OOM failure must hit the best-effort catch too —
            // an uncaught throw on this worker thread would kill the process.
            var r: NeoVapRenderer? = null
            try {
                r = NeoVapRenderer(dummy) {}
                r.awaitInputSurface()
                Log.i("neo_vap", "prewarm: GL pipeline warmed in ${SystemClock.elapsedRealtime() - start}ms")
            } catch (t: Throwable) {
                // swallow — warm is best-effort; log so a silent failure is visible
                Log.w("neo_vap", "prewarm skipped (GL warm failed): ${t.message}")
            } finally {
                r?.release()
            }
        }.apply { name = "neo_vap_prewarm" }.start()
    }

    private fun player(call: MethodCall): NeoVapPlayer? = players[textureId(call)]

    private fun textureId(call: MethodCall): Long =
        call.argument<Number>("textureId")!!.toLong()

    private fun MethodCall.arg(key: String): String = argument<String>(key)!!

    private fun emit(id: Long, event: String, message: String?) = main.post {
        eventSink?.success(mapOf("textureId" to id, "event" to event, "message" to message))
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        players.values.forEach { it.dispose() }
        players.clear()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
    }

    companion object {
        // Process-wide: prewarm the GL pipeline at most once, even across engine
        // re-attach. Best-effort, so it's never reset on failure (no retry storm).
        @Volatile
        private var warmed = false
    }
}
